tensorflow/contrib/avro/ops/parse_avro_record.cc

```diff
@@ -0,0 +1,2058 @@
+                                 // length)
+  std::vector<tensorflow::DataType> sparse_types;  // Types for sparse features
+  std::vector<DenseInformation> dense_infos;  // Information about each dense
+                                              // tensor
+ private:
+  tensorflow::Status FinishInit();  // for context-independent parts of Init.
+};
+
+// As boiler plate I used tensorflow/core/util/example_proto_helper.cc and
+// therein "ParseSingleExampleAttrs" and
+Status CheckValidType(const DataType& dtype) {
+  switch (dtype) {
+    case DT_BOOL:
+    case DT_INT32:
+    case DT_INT64:
+    case DT_FLOAT:
+    case DT_DOUBLE:
+    case DT_STRING:
+      return Status::OK();
+    default:
+      return errors::InvalidArgument("Received input dtype: ",
+                                     DataTypeString(dtype));
+  }
+}
+
+Status CheckDenseShapeToBeDefined(
+    const std::vector<PartialTensorShape>& dense_shapes) {
+  for (int i = 0; i < dense_shapes.size(); ++i) {
+    const PartialTensorShape& dense_shape = dense_shapes[i];
+    bool shape_ok = true;
+    if (dense_shape.dims() == -1) {
+      shape_ok = false;
+    } else {
+      for (int d = 1; d < dense_shape.dims() && shape_ok; ++d) {
+        if (dense_shape.dim_size(d) == -1) {
+          shape_ok = false;
+        }
+      }
+    }
+    if (!shape_ok) {
+      return errors::InvalidArgument(
+          "dense_shapes[", i,
+          "] has unknown rank or unknown inner dimensions: ",
+          dense_shape.DebugString());
+    }
+  }
+  return Status::OK();
+}
+
+// Finishes the initialization for the attributes, which essentially checks that
+// the attributes have the correct values.
+//
+// returns OK if all attributes are valid; otherwise false.
+Status ParseAvroAttrs::FinishInit() {
+  if (static_cast<size_t>(num_sparse) != sparse_types.size()) {
+    return errors::InvalidArgument("len(sparse_keys) != len(sparse_types)");
+  }
+  if (static_cast<size_t>(num_dense) != dense_infos.size()) {
+    return errors::InvalidArgument("len(dense_keys) != len(dense_infos)");
+  }
+  if (num_dense > std::numeric_limits<int32>::max()) {
+    return errors::InvalidArgument("num_dense_ too large");
+  }
+  for (const DenseInformation& dense_info : dense_infos) {
+    TF_RETURN_IF_ERROR(CheckValidType(dense_info.type));
+  }
+  for (const DataType& type : sparse_types) {
+    TF_RETURN_IF_ERROR(CheckValidType(type));
+  }
+  return Status::OK();
+}
+
+// Register the parse function when building the shared library
+// For the op I used as boiler plate: tensorflow/core/ops/parsing_ops.cc and
+// there 'ParseExample'
+// For the op kernel I used as boiler plate:
+// tensorflow/core/kernels/example_parsing_ops.cc and there 'ExampleParserOp'
+// For the compute method I used as boiler plate:
+// tensorflow/core/util/example_proto_fast_parsing.cc and there the
+//   method 'FastParseExample'
+
+REGISTER_OP("ParseAvroRecord")
```

Comment:
Did we discuss combining this with the record reader at some point? Can you add a TODO to combine these? (Unless there's a good reason they're separate? It seems like a waste to reserialize and then deserialize each record between the ops.)

Reply:
Or I guess if you're feeling ambitious you could even make that change now. I think it would make the user experience better and possibly reduce the overall amount of code. I'm fine keeping it as-is for now to get this in though.

Reply:
When I started this work 6 mo ago, the expected output type for the dataset was a tf.tensor with string type.  But looking here: https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/data/ops/dataset_ops.py  this interface has changed and would make it possible to merge the dataset with the parse method.  However, I notice that most existing dataset examples split the load/parse step into two -- often to be able to swap batch/map.  Do you know about the performance impacts the merge of load/parse into one op would have?

Reply:
Can you link to a split load/parse example? I'm not super familiar with the other dataset ops, but at least here I think we'll see better performance by only deserializing and parsing each record once. CsvDataset (https://github.com/tensorflow/tensorflow/blob/master/tensorflow/contrib/data/python/ops/readers.py#L512) is an example of the type of combined read/parse op I'm imagining.

Reply:
I agree with @skye. Currently, it seems `AvroRecordDataset` will return string tensors (each row is a string tensor). Then, `parse_avro_record` applies the filters and returns typed tensors for each of the fields in the records. Seems like `AvroRecordDataset` could be initialized with the filters and just directly return the typed tensors.

Making this change would require significant refactoring, so I'm fine doing this as a follow-up if performance becomes an issue.

Reply:
I've looked up more details; note make_csv_dataset in https://github.com/tensorflow/tensorflow/blob/master/tensorflow/contrib/data/python/ops/readers.py#L512 is marked deprecated.

But I found the recent implementations and hooks here https://github.com/tensorflow/tensorflow/blob/c7bb3c3d65e4e064d53630d4b524522eed6f3f44/tensorflow/contrib/data/python/ops/readers.py.

Note, that the API for Dataset (and now DatasetSource) expects as return type a list of tensors (not a map). I'll adopt the same pattern as for the CsvDataset with the following adjustments:

The AvroDataset returns a list of tensors ordered by
sparse_indices, sparse_values, sparse_shapes, dense_values
In addition, I'll add a method that returns all the keys, which includes the original keys and auxiliary keys to handle sparse data. However, we don't want the user depend on this interface. I'm thinking of not exposing this method publicly at all.

In addition, I add a method make_avro_dataset; which will include assembly into sparse tensors and mapping of keys to tensors, batching etc.  This is the interface we want users depend on.

Reply:
I'll also add a static helper method to the avro dataset that can build a map of keys to tensors (just like before).  This method will take in the list of records together with the list of keys and assemble them together. As described above I don't plan to expose this interface publicly as well; rather it will be used in make_avro_dataset.

Reply:
`make_csv_dataset` and `CsvDataset` seem to have moved to [data/experimental/ops/readers.py](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/data/experimental/ops/readers.py).

I like the idea of adding a `make_avro_dataset` method which will be what users use to create AvroDatasets. I think this method should support returning a dataset where each element is a (features, labels) tuple if the label column name is provided (see [make_csv_dataset](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/data/experimental/ops/readers.py#L310)), so that the dataset can be used directly with the Estimator train/evaluate API.

Another idea to consider is to have the `make_avro_dataset` method take in a `parser_fn` that can transform each record in the dataset. Perhaps you could just pass `parse_avro_record` as the `parser_fn`. See [make_tf_record_dataset](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/data/experimental/ops/readers.py#L218) for inspiration.

I think what should be returned to users calling the `make_avro_dataset` method is a dataset of (optionally) batched (features, labels (optional)) elements. The SparseFeature / SparseTensor, key-mapping stuff, and sparse tensor assembly stuff should all be hidden from users.

Reply:
Makes sense. I guess it doesn't make sense to pass in a `parser_fn` to `make_avro_dataset` since you want to push down the parsing/feature transformation to a lower level. Passing in features should be sufficient; internally, you can then do the needed column projections/transformations to extract/flatten just the requested features.

Reply:
A few suggestions:
* Change `filenames` to `file_pattern` to align with the naming in `make_csv_dataset` and `make_tf_record_dataset`.
* Should we add an optional `label_name` field to support returning elements in the `(features, label)` structure expected by the Estimator APIs?
* Change `num_parallel_calls` to `num_parallel_reads` to align with naming in `make_csv_dataset` and `make_tf_record_dataset`
* Should we use a default parallelism of 1? In some cases, order might matter, so 1 seems like a safer default. Users can always use `Dataset.shuffle()` if they want to perturb the order.

Otherwise, this looks good to me.

Reply:
Refactored `filenames` -> `file_pattern` and `num_parallel_calls` -> `num_parallel_reads`
Changed default parallelism to 1
I'll add the optional `label_name` and logic to the `make_avro_dataset` method but not the `avro_datataset` class (won't be public anyways).